core/optaplanner-constraint-streams/src/main/java/org/optaplanner/constraint/streams/bavet/bi/GroupBiToBiNode.java

```diff
@@ -0,0 +1,163 @@
+            BiConstraintCollector<OldA, OldB, ResultContainer_, B> collector,
+            Consumer<BiTuple<A, B>> nextNodesInsert, Consumer<BiTuple<A, B>> nextNodesRetract) {
+        this.groupKeyMapping = groupKeyMapping;
+        supplier = collector.supplier();
+        accumulator = collector.accumulator();
+        finisher = collector.finisher();
+        this.nextNodesInsert = nextNodesInsert;
+        this.nextNodesRetract = nextNodesRetract;
+        groupMap = new HashMap<>(1000);
+        groupPartMap = new HashMap<>(1000);
+        dirtyGroupQueue = new ArrayDeque<>(1000);
+    }
+
+    private final class Group {
+        A groupKey;
+        ResultContainer_ resultContainer;
+        int parentCount = 0;
+        boolean dirty = false;
+        boolean dying = false;
+        BiTuple<A, B> tupleAB = null;
+
+        public Group(A groupKey, ResultContainer_ resultContainer) {
+            this.groupKey = groupKey;
+            this.resultContainer = resultContainer;
+        }
+    }
+
+    private final class GroupPart {
+        Group group;
+        Runnable undoAccumulator;
+
+        public GroupPart(Group group, Runnable undoAccumulator) {
+            this.group = group;
+            this.undoAccumulator = undoAccumulator;
+        }
+    }
+
+    public void insertAB(BiTuple<OldA, OldB> tupleOldAB) {
+        A groupKey = groupKeyMapping.apply(tupleOldAB.factA, tupleOldAB.factB);
+        Group group = groupMap.computeIfAbsent(groupKey, k -> new Group(groupKey, supplier.get()));
+        group.parentCount++;
+
+        Runnable undoAccumulator = accumulator.apply(group.resultContainer, tupleOldAB.factA, tupleOldAB.factB);
+        GroupPart groupPart = new GroupPart(group, undoAccumulator);
+        GroupPart old = groupPartMap.put(tupleOldAB, groupPart);
+        if (old != null) {
+            throw new IllegalStateException("Impossible state: the tuple for the fact ("
+                    + tupleOldAB.factA + ", " + tupleOldAB.factB + ") was already added in the groupPartMap.");
+        }
+        if (!group.dirty) {
+            group.dirty = true;
+            dirtyGroupQueue.add(group);
+        }
+    }
+
+    public void retractAB(BiTuple<OldA, OldB> tupleOldAB) {
+        GroupPart groupPart = groupPartMap.remove(tupleOldAB);
+        if (groupPart == null) {
+            // No fail fast if null because we don't track which tuples made it through the filter predicate(s)
+            return;
+        }
+        Group group = groupPart.group;
+        group.parentCount--;
+        groupPart.undoAccumulator.run();
+        if (group.parentCount == 0) {
+            Group old = groupMap.remove(group.groupKey);
+            if (old == null) {
+                throw new IllegalStateException("Impossible state: the group for the groupKey ("
+                        + group.groupKey + ") doesn't exist in the groupMap.");
+            }
+            group.dying = true;
```

Comment:
This bears some similarity to tuple state. Maybe it should be an enum too?

Reply:
Maybe, I am not sure - but I am going to rework grouping further and I will consider it then. Thanks.